### PR TITLE
Add stacktrace to the exception response

### LINKED ISSUE #1151: Error handler should not capture exceptions in DEBUG mode
## Description
Based on the [conclusion from the discussion](https://github.com/centerofci/mathesar/discussions/1150).

Improve backend's debugability via two changes:
1. when backend is in debug-mode, add a pretty-printed stacktrace alongside the current error JSON;
2. stop wrapping unhandled exceptions in 4999 errors.

For some context, we're catching backend exceptions and wrapping them in custom errors so as to provide a uniform error interface to API clients. As it is now it has detrimental effects on debugging, because the resulting error messages don't include stacktraces. Above changes aim to address that.

### PR DESCRIPTION
**Pull request by IyadElwy**
Added stacktrace in exception_handler so when a developer gets an error in debug mode, they will additionally get the stacktrace which will lead to an easier debugging experience

Fixes #1151

**Technical details**

**Screenshots**

## Checklist
 
- [ ] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [ ] My pull request targets the `master` branch of the repository
- [ ] My commit messages follow [best practices][best_practices].
- [ ] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [ ] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: mathesar/exception_handlers.py [hunk 1]
```diff
@@ -80,6 +81,7 @@ def mathesar_exception_handler(exc, context):
                 response_data['code'] = error_code
                 response_data['message'] = error_message
                 response_data['details'] = {'exception': force_str(exc)}
+                response_data['stacktrace'] = reformat_stacktrace(traceback.format_exc())
```
Comment (silentninja) [inline mathesar/exception_handlers.py:84]: We should also be able to show the stacktrace even for captured errors that are properly formatted(pretty errors).  Please take a look at line https://github.com/centerofci/mathesar/blob/34d6eb3882c47ff4f43ff346e85c3eac6f0c07e2/mathesar/exception_handlers.py#L62

The response body of those errors should also contain the stacktrace.

**Review by silentninja — CHANGES_REQUESTED** [review]
Nice work! @IyadElwy. I got a few suggestions to improve the PR
- Please make sure the stacktrace is returned only if the `MATHESAR_MODE` is set to `DEVELOPMENT`.
- Please add a test case to demonstrate how the error response would look like

File: mathesar/exception_handlers.py (hunk 1 above)
Comment (IyadElwy) [reply mathesar/exception_handlers.py:84]: Thanks, I'll get on it.

**Comment by kgodey** [comment]
@IyadElwy please replace the `[issue number]` in the PR description with the GitHub issue number, this will link the issue to the PR and make discovering the PR easier. It will also close the issue automatically when this gets merged.

**Comment by IyadElwy** [comment]
> @IyadElwy please replace the `[issue number]` in the PR description with the GitHub issue number, this will link the issue to the PR and make discovering the PR easier. It will also close the issue automatically when this gets merged.

Done.

**Comment by IyadElwy** [comment]
@silentninja I've made the necessary changes and wrote a test case. But I'm not sure where the test case should go. Is there an existing test file for exceptions or should I create one? I figured if there would be one it would be in `mathesar/tests/api` but I'm not sure.

**Comment by IyadElwy** [comment]
I've updated the PR with the latest changes. For the test case, I created a new file `mathesar/tests/api/test_custom_exceptions.py`. If there are any changes I should make please let me know :)

**Review by silentninja — APPROVED** [review]

**Pull request merged**